Code/GraphMol/FindRings.cpp

```diff
@@ -1105,105 +1099,84 @@ int symmetrizeSSSR(ROMol &mol, VECT_INT_VECT &res) {
-        std::cerr<<std::endl;
-        std::cerr<<"------"<<std::endl;
-#endif
-    ssiz = rdcast<unsigned int>(sr.size());
-    INT_VECT exrid;
-    exrid.push_back(srid);
-    Union(bsrs, nunion, &exrid);
-    for (eid = 0; eid < next; eid++) {
-      // if we already added this ring continue
-      // FIX: if the ring has already been added,it probably shouldn't be
-      // in the list at all?  Is this perhaps the most efficient way?
-      if (std::find(symids.begin(), symids.end(), static_cast<int>(eid)) !=
-          symids.end()) {
+  VECT_INT_VECT bondsssrs;
+  RingUtils::convertToBonds(sssrs, bondsssrs, mol);
+
+  //
+  // For each "extra" ring, figure out if it could replace a single
+  // ring in the SSSR. A ring could be swapped out if:
+  //
+  // * They are the same size
+  // * The replacement doesn't remove any bonds from the union of the bonds
+  //   in the SSSR.
+  //
+  // The latter can be checked by determining if the SSSR ring is the unique
+  // provider of any ring bond. If it is, the replacement ring must also
+  // provide that bond.
+  //
+  // May miss extra rings that would need to swap two (or three...) rings
+  // to be included.
+
+  // counts of each bond
+  std::unordered_map<int, int> bondCounts;
+  for (const auto& r: bondsssrs) {
+    for (const auto& b: r) {
+      bondCounts[b] += 1;
+    }
+  }
+
+  INT_VECT extraRing;
+  for (auto& extraAtomRing: extras) {
+    RingUtils::convertToBonds(extraAtomRing, extraRing, mol);
+    for (auto& ring: bondsssrs) {
+      if (ring.size() != extraRing.size()) {
         continue;
       }
-      exr = bextra[eid];
-#if 0
-          std::cerr<<"      "<<eid<<": ";
-          std::copy(extras[eid].begin(),extras[eid].end(),std::ostream_iterator<int>(std::cerr," "));
-          std::cerr<<"  | ";
-          std::copy(exr.begin(),exr.end(),std::ostream_iterator<int>(std::cerr," "));
-          std::cerr<<std::endl;
-#endif
-      if (ssiz == exr.size()) {
-        // std::cerr<<"        possible"<<std::endl;
-        INT_VECT eunion;
-        Union(nunion, exr, eunion);
-#if 0
-            std::cerr<<"           munion: ";
-            std::copy(munion.begin(),munion.end(),std::ostream_iterator<int>(std::cerr," "));
-            std::cerr<<std::endl;
-            std::cerr<<"           eunion: ";
-            std::copy(eunion.begin(),eunion.end(),std::ostream_iterator<int>(std::cerr," "));
-            std::cerr<<std::endl;
-#endif
-        // now check if the eunion is same as the original union from the SSSRs
-        if (eunion.size() == munion.size()) {
-          // we found a symmetric ring
-          symids.push_back(eid);
-          // std::cerr<<"        keep!"<<std::endl;
+
+      // If `ring` is the only provider of some bond, extraRing must also
+      // provide that bond.
+      bool shareABond = false;
+      bool foundAMatch = true;
+      for (auto& bondID: ring) {
+        auto position = find(extraRing.begin(), extraRing.end(), bondID);
```

Comment:
Assuming that you switch to having `bondCounts` be an `std::vector`, I think the following is quicker:
```suggestion
       if(bondCounts[bondID] == 1){
         foundAMatch = false;
         break;
       }
       if(!shareABond){
          auto position = find(extraRing.begin(), extraRing.end(), bondID);
          if (position != extraRing.end()) {
              shareABond = true;
          }
       }
```

Reply:
That doesn't quite work, because bondCounts[bondID] == 1 is OK if extraRing has that bond.

We could skip the std::find for (I'd estimate about) 1/4 of the loops for which there is a match  by doing:
```
      // If `ring` is the only provider of some bond, extraRing must also
      // provide that bond.
      bool shareABond = false;
      bool foundAMatch = true;
      for (auto& bondID: ring) {
        const int bondCount = bondCounts[bondID];
        if (bondCount == 1 || !shareABond) {
          auto position = find(extraRing.begin(), extraRing.end(), bondID);
          if (position != extraRing.end()) {
            shareABond = true;
          } else if (bondCount == 1) {
            // 1 means `ring` is the only ring in the SSSR to provide this bond,
            // and extraRing did not provide it (so extraRing is not an
            // acceptable substitution in the SSSR for ring)
            foundAMatch = false;
            break;
          }
        }
      }
```

This doesn't seem to cause a difference in speed that I can measure, because (in slow systems) most rings don't share a bond so the std::find must be done for every atom in the ring. For instance, in the ~1000 ring nanotube example - only 6 SSSR rings will share a bond, so every bond must be searched in the remaining 994 rings.

I'll see if I can come up with some version of this check that is clearer to read.

(We could probably get more speed by processing the SSSR before the extraRing search. We'd
sort the SSSR rings by size and keep only the ring bonds that are unique to each ring. We could also store the bonds in sorted order to improve the intersection tests. But I think that would be additional complexity, and probably not worth it since SSSR discovery is about 1000x slower than symmetrization)

Reply:
I changed the variable `foundAMatch` to be `replacesAllUniqueBonds`.